bindings/elixir/native/src/ckzg_wrap.c

```diff
@@ -0,0 +1,839 @@
+        }
+
+        if (current_b.size != BYTES_PER_CELL) {
+            enif_free(cell_indices);
+            enif_free(cells);
+            return make_error(env, ckzg_atoms.invalid_cell_length);
+        }
+
+        memcpy(&cells[i], current_b.data, BYTES_PER_CELL);
+    }
+
+    Cell *recovered_cells = enif_alloc(CELLS_PER_EXT_BLOB * BYTES_PER_CELL);
+    if (recovered_cells == NULL) {
+        enif_free(cell_indices);
+        enif_free(cells);
+        return make_error(env, ckzg_atoms.out_of_memory);
+    }
+
+    KZGProof *recovered_proofs = enif_alloc(CELLS_PER_EXT_BLOB * BYTES_PER_PROOF);
+    if (recovered_proofs == NULL) {
+        enif_free(cell_indices);
+        enif_free(cells);
+        enif_free(recovered_cells);
+        return make_error(env, ckzg_atoms.out_of_memory);
+    }
+
+    C_KZG_RET ret = recover_cells_and_kzg_proofs(
+        recovered_cells, recovered_proofs, cell_indices, cells, cells_len, settings
+    );
+    enif_free(cell_indices);
+    enif_free(cells);
+
+    if (ret != C_KZG_OK) {
+        enif_free(recovered_cells);
+        enif_free(recovered_proofs);
+        return make_kzg_error(env, ret);
+    }
+
+    ERL_NIF_TERM *cells_list = enif_alloc(sizeof(ERL_NIF_TERM) * CELLS_PER_EXT_BLOB);
+    if (cells_list == NULL) {
+        enif_free(recovered_cells);
+        enif_free(recovered_proofs);
+        return make_error(env, ckzg_atoms.out_of_memory);
+    }
+
+    ERL_NIF_TERM *proofs_list = enif_alloc(sizeof(ERL_NIF_TERM) * CELLS_PER_EXT_BLOB);
+    if (proofs_list == NULL) {
+        enif_free(recovered_cells);
+        enif_free(recovered_proofs);
+        enif_free(cells_list);
+        return make_error(env, ckzg_atoms.out_of_memory);
+    }
+
+    ErlNifBinary proof_bytes;
+    ErlNifBinary cell_bytes;
+    for (int i = 0; i < CELLS_PER_EXT_BLOB; i++) {
+        if (!enif_alloc_binary(BYTES_PER_CELL, &cell_bytes)) {
+            enif_free(recovered_cells);
+            enif_free(recovered_proofs);
+            enif_free(cells_list);
+            return make_error(env, ckzg_atoms.out_of_memory);
+        }
+
+        memcpy(cell_bytes.data, &recovered_cells[i], BYTES_PER_CELL);
+        cells_list[i] = enif_make_binary(env, &cell_bytes);
+
+        if (!enif_alloc_binary(BYTES_PER_PROOF, &proof_bytes)) {
+            enif_free(recovered_cells);
+            enif_free(recovered_proofs);
+            enif_free(cells_list);
+            return make_error(env, ckzg_atoms.out_of_memory);
+        }
+
+        memcpy(proof_bytes.data, &recovered_proofs[i], BYTES_PER_PROOF);
+        proofs_list[i] = enif_make_binary(env, &proof_bytes);
+    }
+
+    ERL_NIF_TERM cells_term = enif_make_list_from_array(env, cells_list, CELLS_PER_EXT_BLOB);
+    ERL_NIF_TERM proofs_term = enif_make_list_from_array(env, proofs_list, CELLS_PER_EXT_BLOB);
+
+    enif_free(recovered_cells);
+    enif_free(recovered_proofs);
+    enif_free(proofs_list);
+    enif_free(cells_list);
```

Comment:
Yeah haha, these frees are getting wild. I would prefer the `goto` style.

Reply:
yeah.. i have a stashed commit because whilst im not a fan of using `goto`, there is a limit of how many `free()`s you can do haha